### DISCUSSION TITLE
Enable Index to be relocatable

### DISCUSSION THREAD

**Discussion by rowillia** [microsoft/pylance-release#4644] [Enhancement]
To better facilitate ephemeral developer environments in the cloud (like devcontainers) we want to minimize the amount of time it takes from a developer launching the container to when they're writing code, and for large projects indexing can take a while.

We'd like to be able to build an index during CI so developers can pull that index down on start and get started right away (similar to Jetbrain's shared indexes: https://jetbrains.com/help/pycharm/shared-indexes.html#shared-python-packages).  Is that possible in Pylance?

Thanks!

**Comment by rchiodo** [comment]
Thanks for the suggestion. It is probably feasible but not sure we'd want to commit to having this shareable. Right now, it's written to disk next to the extension. At least on the desktop version. In the browser version it's written to an [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) so it's not shareable. But we might end up using an IndexedDB for node too.

Really the root cause of the problem is that indexing takes too long?

**Comment by rchiodo** [comment]
The shared index that Pycharm has is an interesting idea. Instead of you building the indexes we'd do so and then pull them down on startup. That sounds more palatable to me because we wouldn't be exposing how we store the data to anybody. Having an index at all is really an implementation detail.

**Comment by rchiodo** [comment]
Of course, as a temporary workaround, you could just copy the indexes manually :).

They end up here:

![image](https://github.com/microsoft/pylance-release/assets/19672699/48701446-709b-4721-baf3-d6d8c4f41a7c)

Nevermind, it looks like the index has the path inside of the data, so just copying them is not likely to work.

**Comment by rchiodo** [comment]
@rowillia do you have any data on how long indexing is taking your scenario? It should output this in the log like so:

```
2023-07-14 10:22:27.294 [info] (14096) [IDX(2)] index packages c:\Users\rchiodo\source\testing\Testing_Pyright [found 6645 in 1 exec envs] (1043ms)
2023-07-14 10:22:27.295 [info] [Info  - 10:22:27 AM] (14096) indexed(2) 424 files over 1 exec env
2023-07-14 10:22:27.434 [info] [Info  - 10:22:27 AM] (14096) Indexing finished(2).
```

**Comment by rowillia** [comment]
@rchiodo Thanks so much for your responses!!  In our case it takes ~15 seconds on start, partly because it's a large monorepo, partly because the developer's home directory is mounted over NFS.  It's not the end of the world, but it's jarring/annoying enough that people notice.

**Comment by rchiodo** [comment]
Do you guys have the `python.analysis.persistAllIndices` setting set to true? That should speed up indexing significantly, at least on second open.

**Comment by rowillia** [comment]
@rchiodo we don't!  Is there documentation for it?  I'll give it a shot now!

**Comment by rchiodo** [comment]
Oh I thought we had fully released it. It's still experimental at the moment, but should work:

```
                "python.analysis.persistAllIndices": {
                    "type": "boolean",
                    "default": false,
                    "description": "Indices for all third party libraries will be persisted to disk.",
                    "scope": "resource",
                    "tags": [
                        "experimental"
                    ]
                },
```

The only documentation is in the package.json.

**Comment by judej** [comment]
Converting this issue to a discussion to learn how useful this feature would be for others.
